[batch] add resources to billing

jigold opened the pull request:
I think this can go in instead of #8730. I ran dev deploy with master and then didn't delete the database and ran the tests with the new version. The billing UI page reported the correct values. I also ran the new version with the check functions in the background and got no errors.

I can probably double check the UI batches cost are correct, but let's wait until we're happy with the code before I do anymore testing.

danking commented:
Sorry, I know I'm late on this. I will make it a priority tomorrow.

danking commented on batch/batch/worker_config.py:
```diff
@@ -0,0 +1,131 @@
+        resources = []
+
+        preemptible = 'preemptible' if self.preemptible else 'nonpreemptible'
+        worker_fraction = cpu_in_mcpu / (self.cores * 1000)
```
This calculation should happen entirely with integral arithmetic. We want to get to 1024ths of an instance. We have used mcpu and total cores.

Thanks to your powers-of-two changes, mcpu should always divide cores * 1000 (mcpu/1000 must be a smaller power of two so it will always divide cores)
```
(cores * 1000) // mcpu
```
This tells us the nth of an instance we have. For 250 mCPU of a 16 core instance this is 64, so we have one-sixty-fourth of an instance. Now we just divide 1024 by our nth to get our fraction of 1024:
```
1024 // ((cores * 1000) // mcpu)
```
This works out to 16-1024ths. A few more worked examples:
```
In [116]: 1024 // ((256 * 1000) // 250)                                                                                                                                                                                                                                                                                                                
Out[116]: 1

In [117]: 1024 // ((128 * 1000) // 250)                                                                                                                                                                                                                                                                                                                
Out[117]: 2

In [118]: 1024 // ((8 * 1000) // 8000)                                                                                                                                                                                                                                                                                                                 
Out[118]: 1024
```

danking commented on batch/batch/worker_config.py:
```diff
@@ -0,0 +1,131 @@
+                          'quantity': math.ceil(memory_in_bytes / 1024 / 1024)})
+
+        resources.append({'name': f'boot-disk/{self.boot_disk_type}/1',
+                          'quantity': math.ceil(self.boot_disk_size_gb * worker_fraction * 1024)})
```
Define `worker_fraction_in_1024ths` as `1024 // ((cores * 1000) // mcpu)`, then this is just `self.boot_disk_size_gb * worker_fraction_in_1024ths`.

danking replied:
And no ceil, right?

danking commented on batch/batch/resources.py:
```diff
@@ -0,0 +1,54 @@
+            # usage is in units of msec-mb
+            # average number of days per month = 365.25 / 12 = 30.4375
+            avg_n_days_per_month = 30.4375
+            cost = rate / avg_n_days_per_month / 24 / 3600 * (usage * 0.001 * 0.001)
```
One of these should be `/ 1024` right?

jigold replied:
Good catch!

danking commented on batch/batch/worker_config.py:
```diff
@@ -0,0 +1,131 @@
+                          'quantity': cpu_in_mcpu})
+
+        resources.append({'name': f'memory/{self.instance_family}-{self.instance_type}-{preemptible}/1',
+                          'quantity': math.ceil(memory_in_bytes / 1024 / 1024)})
```
these should be 1000.

danking replied:
All of our machines' memory in bytes should be cleanly divided by 1_000_000, right? We should assert that and use `memory_in_bytes // 1_000_000`.

danking replied:
no ceil.

jigold commented on batch/batch/utils.py:
```diff
@@ -83,7 +101,7 @@ def worker_memory_per_core_gb(worker_type):
 def worker_memory_per_core_bytes(worker_type):
     m = worker_memory_per_core_gb(worker_type)
-    return int(m * 1000**3)  # GCE memory/core are in GB not GiB
+    return int(m * 1024**3)
```
I was trying to figure out why the old method of computing cost wasn't aligning up with the new way. So one hypothesis was I was using powers of 1000 instead of 1024 and must have forgotten to change it back.

I went back and tried to figure out what the right answer is and I think it should be powers of 1000 for memory, but I'd appreciate you double checking. I'll try and take a look at disk separately.

```
jigold@batch-worker-jigold-5utdq:~$ free -b
              total        used        free      shared  buff/cache   available
Mem:     3869945856   427466752   416616448     1286144  3025862656  3213860864
Swap:             0           0           0
```

If you divide 3869945856 / 1024**3, you get 3.6Gi
If you divide 3869945856 / 1000**3, you get 3.87GB

The RAM we're supposed to have available is 3.75GB. I couldn't find any documentation anywhere whether that was supposed to be in GB or Gi. So I assumed it had to be in GB based on above.

I tried a different way to get this answer:

```
jigold@batch-worker-jigold-5utdq:~$ grep 'MemTotal' /proc/meminfo
MemTotal:        3779244 kB
```

3779244 / 1000 / 1000 = 3.78
3779244 / 1024 / 1024 = 3.60

The --si option in free is supposed to print in powers of 1000

```
jigold@batch-worker-jigold-5utdq:~$ free --si
              total        used        free      shared  buff/cache   available
Mem:        3779244      413400      410276        1256     2955568     3142584
Swap:             0           0           0
```

So this is where I got the powers of 1000 is probably the correct answer and we can say the instance has 3.75GB of RAM.

jigold replied:
I confirmed disk size is in GiB.

```
jigold@batch-worker-jigold-5utdq:~$ sudo fdisk -l | grep Disk
Disk /dev/loop0: 66.5 MiB, 69758976 bytes, 136248 sectors
Disk /dev/loop1: 89.1 MiB, 93454336 bytes, 182528 sectors
Disk /dev/loop2: 93.9 MiB, 98484224 bytes, 192352 sectors
Disk /dev/loop3: 55 MiB, 57618432 bytes, 112536 sectors
Disk /dev/loop4: 98.7 MiB, 103432192 bytes, 202016 sectors
Disk /dev/sda: 20 GiB, 21474836480 bytes, 41943040 sectors
Disklabel type: gpt
Disk identifier: EE3550B4-9533-4B5A-8E61-DE881FE8AF66
Disk /dev/nvme0n1: 375 GiB, 402653184000 bytes, 98304000 sectors
```

jigold commented on batch/batch/utils.py:
```diff
@@ -1,14 +1,32 @@
+
+    worker_type = 'standard'
+    worker_cores = 16
+    worker_disk_size_gb = 100
```
There should be no records before 16 core days. I took a look at all the migrations.

jigold commented on batch/batch/driver/main.py:
```diff
@@ -474,6 +481,145 @@ def user_get(d, u, f):
         await asyncio.sleep(0.1)
 
 
+async def check_resource_aggregation(db):
```
It took a long time to get the code correct for these checks. We kept Cotton's incremental check in the codebase. I can see the value both ways. We're never going to run these queries on the full database, so maybe it's best to keep them out. I'll at least keep them while I'm finishing this PR and we can make a final decision later.

jigold commented on batch/batch/globals.py:
```diff
@@ -16,6 +16,7 @@
-STATUS_FORMAT_VERSION = 2
-INSTANCE_VERSION = 3
+BATCH_FORMAT_VERSION = 3
+STATUS_FORMAT_VERSION = 3
```
can you be more specific? We're changing the format of the status written on GCE.

jigold commented on batch/batch/globals.py:
```diff
@@ -16,6 +16,7 @@
-INSTANCE_VERSION = 3
+BATCH_FORMAT_VERSION = 3
+STATUS_FORMAT_VERSION = 3
+INSTANCE_VERSION = 4
```
It's used to know what version of the instances we have in our current worker pool. We have to increment the version because they're sending a slightly different status to the driver.

jigold commented on batch/batch/resources.py:
```diff
@@ -0,0 +1,54 @@
+    'boot-disk/pd-ssd/1': 0.17,  # gb-month
+    'ip-fee/1024/1': 0.004,  # instance-hour
+    'service-fee/1': 0.01  # cpu-hour
+}
```
I thought this was easiest, but I kept waffling back and forth on what the right thing to do is. If we do it in MySQL, then I want an easy way to make sure the instance configuration will generate valid resource types and a UI page to add new resources and rates (won't be able to modify existing entries). It seemed easier to see how the cost was computed in Python where I could do something like `0.17 / avg_n_days_per_month / ...`. But I can see how having this in SQL would be better.

Can we discuss this more? Maybe in person if that's easier?

jigold commented:
I addressed all your comments except making the decision how this should be structured between MySQL and Python. Once we make that decision, then I'll make the changes and double check with dev deploy that the costs are still the same and the resource aggregation is correct. Sound good?

jigold commented:
The one thing I wasn't sure about is whether I got the indexes right for the billing front_end query. I didn't know if the subqueries needed the combined index or they could just filter on format_version

danking commented on batch/batch/batch.py:
```diff
@@ -177,6 +198,20 @@ async def mark_job_started(app, batch_id, job_id, attempt_id, instance, start_ti
     if rv['delta_cores_mcpu'] != 0 and instance.state == 'active':
         instance.adjust_free_cores_in_memory(rv['delta_cores_mcpu'])
 
+    try:
```
the error message should include the attempt id

danking commented on batch/batch/front_end/front_end.py:
```diff
@@ -1119,20 +1138,40 @@ async def parse_error(msg):
+           ON aggregated_batch_resources.batch_id = tmp1.id
+           INNER JOIN resources ON resources.resource = aggregated_batch_resources.resource
+           GROUP BY billing_project, `user`) AS t3
+ON t1.billing_project = t3.billing_project AND t1.`user` = t3.`user`;
```
let's go with
```sql
SELECT
  billing_project,
  `user`,
  CAST(COALESCE(SUM(IF(format_version < 3, msec_mcpu, 0)), 0) AS SIGNED) as msec_mcpu,
  SUM(IF(format_version >= 3, `usage` * rate, NULL)) as cost
FROM batches
LEFT JOIN aggregated_batch_resources
       ON aggregated_batch_resources.batch_id = batches.id
LEFT JOIN resources
        ON resources.resource = aggregated_batch_resources.resource
GROUP BY billing_project, `user`
```

danking commented on batch/batch/front_end/front_end.py:
```diff
@@ -1119,20 +1138,29 @@ async def parse_error(msg):
+    def billing_record_to_dict(record):
+        cost_msec_mcpu = cost_from_msec_mcpu(record['msec_mcpu'])
+        cost_resources = record['cost']
+        record['cost'] = coalesce(cost_msec_mcpu, 0) + coalesce(cost_resources, 0)
```
ok tiny thing, we have two coalesces here but one in the SQL, seems simpler to do it in one or the other. In particular, `cost_msec_mcpu` should always be a number, right? Maybe just remove the coalesce from SQL? `CAST(NULL AS SIGNED)` is still `NULL`, so I think it shouldn't affect the semantics.
